### PR TITLE
fix(codepipeline-actions): cross stack reference causes stack cycle in sources that use CloudWatch Events

### LINKED ISSUE #3087: Cross stack reference caused "cyclic reference" error
**Note: for support questions, please first reference our [documentation](https://docs.aws.amazon.com/cdk/api/latest), then use [Stackoverflow](https://stackoverflow.com/questions/ask?tags=aws-cdk)**. This repository's issues are intended for feature requests and bug reports.

* **I'm submitting a ...**
  - [x] :beetle: bug report
  - [ ] :rocket: feature request
  - [ ] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/awslabs/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **What is the current behavior?**
*If the current behavior is a :beetle:bug:beetle:: Please provide the steps to reproduce*

I created 2 stacks - one for CodeCommit repository, and the other for CodePipeline which takes the previous one as source stage. 

```python
class CodeRepo(core.Stack):
......
    self.repo = codecommit.Repository(
            self,
            id='CodeCommitRepo',
            repository_name=repo_name
        )

class MyPipeline(core.Stack):
......
        def __init__(self, scope: core.Construct, id: str, code_repo: codecommit.IRepository) -> None:
            ......
            # create CodePipeline and its stages.
            pipeline = codepipeline.Pipeline(self, id='CodePipeline', pipeline_name='pipeline', restart_execution_on_update=False)

            source_stage = pipeline.add_stage(stage_name='Source')
            source_stage.add_action(
                codepipeline_actions.CodeCommitSourceAction(
                    action_name='action',
                    repository=code_repo,
                    branch='master',
                    output=codepipeline.Artifact(artifact_name='output'),
                    trigger=codepipeline_actions.CodeCommitTrigger.EVENTS
                )
            )

app = core.App() 
repo_stack = CodeRepo(app, 'repo_stack')
pipeline_stack = MyPipeline(app, 'pipeline_stack', repo_stack.repo)

app.synth()
```

Then I got following error:

```shell
File "./venv/lib/python3.6/site-packages/aws_cdk/core/__init__.py", line 2905, in synth
    return jsii.invoke(self, "synth", [])
  File "./venv/lib/python3.6/site-packages/jsii/_kernel/__init__.py", line 104, in wrapped
    return _recursize_dereference(kernel, fn(kernel, *args, **kwargs))
  File "./venv/lib/python3.6/site-packages/jsii/_kernel/__init__.py", line 258, in invoke
    args=_make_reference_for_native(self, args),
  File "./venv/lib/python3.6/site-packages/jsii/_kernel/providers/process.py", line 346, in invoke
    return self._process.send(request, InvokeResponse)
  File "./venv/lib/python3.6/site-packages/jsii/_kernel/providers/process.py", line 316, in send
    raise JSIIError(resp.error) from JavaScriptError(resp.stack)
jsii.errors.JSIIError: 'pipeline_stack' depends on 'repo_stack' (pipeline_stack/CodePipeline/Role/DefaultPolicy/Resource -> repo_stack/CodeCommitRepo/Resource.Arn). Adding this dependency (repo_stack/CodeCommitRepo/pipelinestackCodePipelineA06C8E1EEventRule/Resource -> pipeline_stack/CodePipeline/Resource.Ref) would create a cyclic reference.
```

An interesting observation is, if i put ```codecommit.Repository(...)```  in the pipeline_stack and refer the repo resource as usual, everything works without cyclic dependency. 

* **What is the expected behavior (or behavior of feature suggested)?**
CDK is expected to handle the dependency issue, as described on [aws doc](https://docs.aws.amazon.com/cdk/latest/guide/aws_construct_lib.html) for "Passing Resources from a Different Stack".

* **What is the motivation / use case for changing the behavior or adding this feature?**

* **Please tell us about your environment:**
  
  - CDK CLI Version: 0.36.0
  - Module Version: 0.36.0
  - OS: [ OSX Mojave ]
  - Language: [ Python ]

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. associated pull-request, stackoverflow, gitter, etc)

### LINKED ISSUE #8042: CodeCommit and CodePipeline in different stacks leads to circular dependency
## :question: General Issue
I want to have a CodeCommit repository in one stack triggering a CodePipeline in another stack, both stacks being deployed to the same region.

When passing the repo as a prop to the stack containing the pipeline, I get a circular dependency error:
`
'RepoStack' depends on 'PipelineStack' (RepoStack -> PipelineStack/Pipeline/Resource.Ref, RepoStack -> PipelineStack/Pipeline/EventsRole/Resource.Arn). Adding this dependency (PipelineStack -> RepoStack/Repository/Resource.Name) would create a cyclic reference.
`

My code looks as follows:

```js
const repoStack = new RepoStack(app, 'RepoStack', {
  env: envDevOps,
});
new PipelineStack(app, 'PipelineStack', {
  env: envDevOps,
  repo: repoStack.repo,
});
```

```js
export class RepoStack extends cdk.Stack {
  public readonly repo: codecommit.Repository;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.repo = new codecommit.Repository(this, 'Repository', {
      repositoryName: 'MyRepo',
    });
  }
}
```

```js
export class PipelineStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props: PipelineStackProps) {
    super(scope, id, props);

    const sourceAction = new codepipeline_actions.CodeCommitSourceAction({
      actionName: 'CodeCommit_Source',
      repository: props.repo,
      output: sourceOutput,
    });
    
    // ...
  }
}
```

### The Question

How can I have a CodeCommit repository in one stack triggering a CodePipeline in another stack without introducing a circular dependency (both in the same account/region)?

When the `RepoStack` is deployed to another account, everything works fine. I suspect that, in this case, the cross-account support stacks do the magic.
Still, it comes to a surprise that support for cross-references across accounts is better than support for cross-references within the same account.

I think this problem can be generalized to many scenarios with circular dependencies.

### Environment

  - **CDK CLI Version:** 1.39.0 (build 5d727c1)
  - **OS:** MacOS Catalina
  - **Language:** all

### LINKED ISSUE #10896: [aws-codepipeline] Cyclic reference error when importing an S3 bucket in a code pipeline defined in a different stack
## :question: General Issue
Hi- I am using CDK to create a code pipeline with an S3 bucket as its source. The S3 bucket and code pipeline are defined in different stacks and per the AWS documentation we can access resources in a different stack as long as they are in the same account and AWS region: https://docs.aws.amazon.com/cdk/latest/guide/resources.html#resource_stack

### The Question

This doesn't seem to be working and I am getting a "Cyclic reference" error. Does this mean the AWS documentation is wrong? Or is there something in particular that's missing to make this work? Sample code below

```ts
        const pipelineSourceActionOutput = new Artifact();
        const pipelineSourceAction = new S3SourceAction({
            actionName: "S3SourceAction",
            bucket: s3Bucket, // Versioned Bucket (and CloudTrail) defined in a different stack
            bucketKey: `${id}/upload.zip`,
            output: pipelineSourceActionOutput,
            trigger: S3Trigger.EVENTS,
            runOrder: 1,
        });
        new Pipeline(this, pipelineId, {
            pipelineName: pipelineId,
            stages: [
                {
                    stageName: "Source",
                    actions: [pipelineSourceAction],
                }
            ],
        });
```

I think this line of code is causing the dependency of my stack containing S3 bucket on the stack that defined code pipeline infra: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-codepipeline-actions/lib/s3/source-action.ts#L120-L124

### Environment

  - **CDK CLI Version:** 1.45.0
  - **Module Version:** 
  - **Node.js Version:** 12.x.31654.0
  - **OS:** OSX Mojave 
  - **Language (Version):** TypeScript

### PR DESCRIPTION
**Pull request by skinny85**
When using a newly-created, CDK-managed resource, such as an S3 Bucket or a CodeCommit Repository,
as the source of a CodePipeline, when it's in a different Stack than the pipeline
(but in the same environment as it),
and we use CloudWatch Events to trigger the pipeline from the source,
that would result in a cycle:

1. Because the Event Rule is created in the source Stack, that Stack needs the name of the pipeline to trigger from the Rule, and also the Role to use for the trigger, which is created in the target (in this case, the pipeline) Stack. So, the source Stack depends on the pipeline Stack.
2. The pipeline Stack needs the name of the source resource. So, the pipeline Stack depends on the source Stack.

The only way to break this cycle is to move the Event Rule to the target Stack.
This PR adds an API to the Events module to make it possible for event-creating constructs to make that choice,
and uses that capability in the CodePipeline `CodeCommitSourceAction` and `S3SourceAction`.

Fixes #3087
Fixes #8042
Fixes #10896

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by skinny85** [comment]
@rix0rrr would be curious to get your opinion on whether you like this approach!

File: packages/@aws-cdk/aws-events/lib/on-event-options.ts [hunk 1]
```diff
@@ -39,4 +33,24 @@ export interface OnEventOptions {
...
    */
   readonly eventPattern?: EventPattern;
+
+  /**
+   * The scope to use if the source of the rule and its target are in different Stacks.
+   * This helps dealing with cycles that often arise in these situations.
+   *
+   * @default - none (the main scope will be used, even for cross-stack Events)
+   */
+  readonly scopeIfCrossStack?: Construct;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-events/lib/on-event-options.ts:43]: This name definitely sucks, so I'm very open to suggestions for a better one.

**Comment by skinny85** [comment]
Note that, before this PR is merged, we should make a similar change that we have now in `CodeCommitSourceAction` also in `S3SourceAction`.

File: packages/@aws-cdk/aws-events/lib/on-event-options.ts (hunk 1 above)
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-events/lib/on-event-options.ts:43]: Maybe just crossStackScope?

File: packages/@aws-cdk/aws-events/lib/rule.ts [hunk 2]
```diff
@@ -119,7 +88,7 @@ export class Rule extends Resource implements IRule {
   private readonly _xEnvTargetsAdded = new Set<string>();
 
   constructor(scope: Construct, id: string, props: RuleProps = { }) {
-    super(scope, id, {
+    super(determineRuleScope(scope, props), id, {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-events/lib/rule.ts:91]: Tests for this?

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
So, I think I like this but I'd definitely want to see more until testing and some integration tests.

File: packages/@aws-cdk/aws-events/lib/on-event-options.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-events/lib/on-event-options.ts:43]: OK, I'll update to `crossStackScope`.

File: packages/@aws-cdk/aws-events/lib/rule.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-events/lib/rule.ts:93]: This is verified by every single test that uses any `Rule`, of which there are hundreds, and also the new `"allows using a new Repository from another Stack as a source of the Pipeline, with Events"` test that was added in this PR.

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-events/lib/rule.ts:93]: Fair. I can let that go.

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-events/lib/rule.ts:93]: Can we get an integration test on this, though? It seems like exactly the sort of thing that we'd want integ tests for.

Comment (skinny85) [reply packages/@aws-cdk/aws-events/lib/rule.ts:93]: Sure 🙂.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Just putting this in changes requested so I see when an update has been made. You mentioned that this sam change should be made for `S3SourceAction`. Is that a change you'd like to make or should I go ahead and take that on?

**Comment by skinny85** [comment]
> Just putting this in changes requested so I see when an update has been made. You mentioned that this sam change should be made for `S3SourceAction`. Is that a change you'd like to make or should I go ahead and take that on?

I'll do it. I'll combine with the integration test.

**Comment by skinny85** [comment]
> > Just putting this in changes requested so I see when an update has been made. You mentioned that this sam change should be made for `S3SourceAction`. Is that a change you'd like to make or should I go ahead and take that on?
> 
> I'll do it. I'll combine it with the integration test.

Done in the latest revision. @TheRealAmazonKendra would appreciate another look!

**Review by TheRealAmazonKendra — DISMISSED** [review]
Just one more issue in here. Looks like this isn't using the new [IntegTest construct](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) to run the integration test. Can you please make that update? Besides that, this looks good to go!

**Comment by skinny85** [comment]
> Just one more issue in here. Looks like this isn't using the new [IntegTest construct](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) to run the integration test. Can you please make that update? Besides that, this looks good to go!

Done!

**Review by TheRealAmazonKendra — APPROVED** [review]
Looks great!

**Comment by skinny85** [comment]
Is `main` broken right now...?

```
@aws-cdk/integ-runner: PASS test/runner/snapshot-test-runner.test.js
@aws-cdk/integ-runner: PASS test/runner/integ-test-runner.test.js
@aws-cdk/integ-runner: FAIL test/workers/integ-worker.test.js
@aws-cdk/integ-runner:   â— test runner â€º has snapshot
@aws-cdk/integ-runner:     expect(received).toEqual(expected) // deep equality
@aws-cdk/integ-runner:     - Expected  - 1
@aws-cdk/integ-runner:     + Received  + 6
@aws-cdk/integ-runner:     - Array []
@aws-cdk/integ-runner:     + Array [
@aws-cdk/integ-runner:     +   Object {
@aws-cdk/integ-runner:     +     "discoveryRoot": "test/test-data",
@aws-cdk/integ-runner:     +     "fileName": "test/test-data/xxxxx.test-with-snapshot.js",
@aws-cdk/integ-runner:     +   },
@aws-cdk/integ-runner:     + ]
@aws-cdk/integ-runner:       150 |     ]));
@aws-cdk/integ-runner:       151 |
@aws-cdk/integ-runner:     > 152 |     expect(results).toEqual([]);
@aws-cdk/integ-runner:           |                     ^
@aws-cdk/integ-runner:       153 |   });
@aws-cdk/integ-runner:       154 |
@aws-cdk/integ-runner:       155 |   test('deploy failed', () => {
@aws-cdk/integ-runner:       at Object.<anonymous> (test/workers/integ-worker.test.ts:152:21)
@aws-cdk/integ-runner: PASS test/workers/snapshot-worker.test.js
@aws-cdk/integ-runner: =============================== Coverage summary ===============================
@aws-cdk/integ-runner: Statements   : 81.1% ( 571/704 )
@aws-cdk/integ-runner: Branches     : 71.94% ( 377/524 )
@aws-cdk/integ-runner: Functions    : 81.1% ( 103/127 )
@aws-cdk/integ-runner: Lines        : 81.21% ( 549/676 )
@aws-cdk/integ-runner: ================================================================================
@aws-cdk/integ-runner: Test Suites: 1 failed, 7 passed, 8 total
@aws-cdk/integ-runner: Tests:       1 failed, 63 passed, 64 total
@aws-cdk/integ-runner: Snapshots:   0 total
@aws-cdk/integ-runner: Time:        4.34 s
@aws-cdk/integ-runner: Ran all test suites.
@aws-cdk/integ-runner: Error: /codebuild/output/src162968263/src/github.com/aws/aws-cdk/node_modules/jest/bin/jest.js exited with error code 1
@aws-cdk/integ-runner: Tests failed. Total time (5.3s) | /codebuild/output/src162968263/src/github.com/aws/aws-cdk/node_modules/jest/bin/jest.js (5.3s)
```

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by TheRealAmazonKendra** [comment]
> Is `main` broken right now...?
> 
> ```
> @aws-cdk/integ-runner: PASS test/runner/snapshot-test-runner.test.js
> @aws-cdk/integ-runner: PASS test/runner/integ-test-runner.test.js
> @aws-cdk/integ-runner: FAIL test/workers/integ-worker.test.js
> @aws-cdk/integ-runner:   â—� test runner â€º has snapshot
> @aws-cdk/integ-runner:     expect(received).toEqual(expected) // deep equality
> @aws-cdk/integ-runner:     - Expected  - 1
> @aws-cdk/integ-runner:     + Received  + 6
> @aws-cdk/integ-runner:     - Array []
> @aws-cdk/integ-runner:     + Array [
> @aws-cdk/integ-runner:     +   Object {
> @aws-cdk/integ-runner:     +     "discoveryRoot": "test/test-data",
> @aws-cdk/integ-runner:     +     "fileName": "test/test-data/xxxxx.test-with-snapshot.js",
> @aws-cdk/integ-runner:     +   },
> @aws-cdk/integ-runner:     + ]
> @aws-cdk/integ-runner:       150 |     ]));
> @aws-cdk/integ-runner:       151 |
> @aws-cdk/integ-runner:     > 152 |     expect(results).toEqual([]);
> @aws-cdk/integ-runner:           |                     ^
> @aws-cdk/integ-runner:       153 |   });
> @aws-cdk/integ-runner:       154 |
> @aws-cdk/integ-runner:       155 |   test('deploy failed', () => {
> @aws-cdk/integ-runner:       at Object.<anonymous> (test/workers/integ-worker.test.ts:152:21)
> @aws-cdk/integ-runner: PASS test/workers/snapshot-worker.test.js
> @aws-cdk/integ-runner: =============================== Coverage summary ===============================
> @aws-cdk/integ-runner: Statements   : 81.1% ( 571/704 )
> @aws-cdk/integ-runner: Branches     : 71.94% ( 377/524 )
> @aws-cdk/integ-runner: Functions    : 81.1% ( 103/127 )
> @aws-cdk/integ-runner: Lines        : 81.21% ( 549/676 )
> @aws-cdk/integ-runner: ================================================================================
> @aws-cdk/integ-runner: Test Suites: 1 failed, 7 passed, 8 total
> @aws-cdk/integ-runner: Tests:       1 failed, 63 passed, 64 total
> @aws-cdk/integ-runner: Snapshots:   0 total
> @aws-cdk/integ-runner: Time:        4.34 s
> @aws-cdk/integ-runner: Ran all test suites.
> @aws-cdk/integ-runner: Error: /codebuild/output/src162968263/src/github.com/aws/aws-cdk/node_modules/jest/bin/jest.js exited with error code 1
> @aws-cdk/integ-runner: Tests failed. Total time (5.3s) | /codebuild/output/src162968263/src/github.com/aws/aws-cdk/node_modules/jest/bin/jest.js (5.3s)
> ```

Wut? Gotta be a transient issue. Let's see what the retry does.

**Pull request merged**